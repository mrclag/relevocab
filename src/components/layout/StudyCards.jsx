import React, { useState, useEffect } from 'react';
import firebase from '../../firebase';

import styled from 'styled-components';

import Card from '../Card';

const StudyCards = ({ deck }) => {
  console.log('rerender studycard');
  const [cards, setCards] = useState([
    {
      id: 0,
      eng: 'Deck of Cards',
      foreign: 'Deck of Cards'
    }
  ]);
  const [currentCard, setCurrentCard] = useState({});
  const [currentDeck, setCurrentDeck] = useState([]);

  useEffect(() => {
    firebase
      .firestore()
      .collection(deck.value)
      .orderBy('foreign', 'desc')
      .onSnapshot(snapshot => {
        const newVocab = snapshot.docs.map(doc => ({
          id: doc.id,
          ...doc.data()
        }));

        setCurrentDeck(newVocab);
        setCards(newVocab);
      });
  }, [deck]);

  const getRandomCard = currentCards => {
    let card = currentCards[Math.floor(Math.random() * currentCards.length)];
    return card;
  };

  const updateCard = () => {
    setCurrentCard(getRandomCard(cards));
  };

  const removeCard = () => {
    const filteredCards = cards.filter(card => card.id !== currentCard.id);
    setCards(filteredCards);
    setCurrentCard(getRandomCard(filteredCards));
  };

  const resetDeck = () => {
    setCards(currentDeck);
    setCurrentCard(getRandomCard(currentDeck));
  };

  return (
    <ContentWrapper>
      Cards in deck: {cards.length} / {currentDeck.length}
      {currentCard ? (
        <Card eng={currentCard.eng} foreign={currentCard.foreign} />
      ) : (
        <Card
          eng={<div style={{ color: 'red' }}>End of Deck</div>}
          foreign="End of Deck"
        />
      )}
      <div>
        {cards.length > 0 ? (
          <>
            <CardButton color="red" onClick={updateCard}>
              Again
            </CardButton>
            <CardButton color="green" onClick={() => removeCard()}>
              Good
            </CardButton>
          </>
        ) : (
          <CardButton color="black" onClick={() => resetDeck()}>
            Reset
          </CardButton>
        )}
      </div>
    </ContentWrapper>
  );
};

export default StudyCards;

const ContentWrapper = styled.div`
  display: flex;
  position: absolute;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  top: 200px;
  left: 40%;
`;

const CardButton = styled.button`
  height: 50px;
  width: 75px;
  background-color: ${props => props.color};
  border: none;
  color: white;
  margin: 10px;
`;
