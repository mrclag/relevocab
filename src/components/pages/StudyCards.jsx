import React, { useState, useEffect } from 'react';
import firebase from '../../firebase';
import Card from '../Card';

import {
  ContentWrapper,
  CardButton
} from '../../styles/pages/StudyCards.styles';

const StudyCards = ({ deck }) => {
  console.log('Render StudyCard');
  const [cards, setCards] = useState([
    {
      id: 0,
      eng: 'Deck of Cards',
      foreign: 'Deck of Cards'
    }
  ]);
  const [currentCard, setCurrentCard] = useState({});
  const [currentDeck, setCurrentDeck] = useState([]);

  useEffect(() => {
    firebase
      .firestore()
      .collection(deck.value)
      .orderBy('foreign', 'desc')
      .onSnapshot(snapshot => {
        const newVocab = snapshot.docs.map(doc => ({
          id: doc.id,
          ...doc.data()
        }));
        console.log('Getting StudyCards');
        setCurrentDeck(newVocab);
        setCards(newVocab);
      });
  }, [deck]);

  const getRandomCard = currentCards => {
    let card = currentCards[Math.floor(Math.random() * currentCards.length)];
    return card;
  };

  const updateCard = () => {
    setCurrentCard(getRandomCard(cards));
  };

  const removeCard = () => {
    const filteredCards = cards.filter(card => card.id !== currentCard.id);
    setCards(filteredCards);
    setCurrentCard(getRandomCard(filteredCards));
  };

  const resetDeck = () => {
    setCards(currentDeck);
    setCurrentCard(getRandomCard(currentDeck));
  };

  return (
    <ContentWrapper>
      Cards in deck: {cards.length} / {currentDeck.length}
      {currentCard ? (
        <Card
          eng={currentCard.eng || deck.value}
          foreign={currentCard.foreign}
        />
      ) : (
        <Card
          eng={<div style={{ color: 'red' }}>End of Deck</div>}
          foreign="End of Deck"
        />
      )}
      <div>
        {cards.length > 0 ? (
          currentCard && currentCard.eng ? (
            <>
              <CardButton color="red" onClick={updateCard}>
                Again
              </CardButton>
              <CardButton color="green" onClick={() => removeCard()}>
                Good
              </CardButton>
            </>
          ) : (
            <CardButton color="#107bbd" onClick={() => resetDeck()}>
              Start
            </CardButton>
          )
        ) : (
          <CardButton color="black" onClick={() => resetDeck()}>
            Reset
          </CardButton>
        )}
      </div>
    </ContentWrapper>
  );
};

export default StudyCards;
