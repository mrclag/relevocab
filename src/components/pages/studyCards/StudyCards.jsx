import React, { useState } from 'react';
import { connect } from 'react-redux';

import { ContentWrapper, CardButton } from './StudyCards.styles';

import Card from './Card';

const StudyCards = ({ currentDeck }) => {
  const [cardPile, setCardPile] = useState([]);
  const [currentCard, setCurrentCard] = useState({
    front: currentDeck.title,
    back: currentDeck.title,
    start: true
  });
  const [flipped, setFlipped] = useState(false);

  const updateCard = () => {
    // Remove current card from top of pile, put it at the rear
    const oldCard = cardPile[0];
    const newCardPile = cardPile.slice(1, cardPile.length);
    newCardPile.push(oldCard);
    setCardPile(newCardPile);
    setCurrentCard(cardPile[0]);
    setFlipped(false);
  };

  const removeCard = () => {
    const newCardPile = cardPile.slice(1, cardPile.length);
    setCardPile(newCardPile);
    setCurrentCard(cardPile[0]);
    setFlipped(false);
  };

  const randomizeDeck = () => {
    setCurrentCard({
      front: currentDeck.title,
      back: currentDeck.title,
      start: true
    });
    // converting object of cards to array
    const cards = Object.keys(currentDeck.cards).map(
      (key, i) => currentDeck.cards[key]
    );
    // Fisher shuffle: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm
    for (let i = cards.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [cards[i], cards[j]] = [cards[j], cards[i]];
    }
    setCardPile(cards);
    setCurrentCard(cards[0]);
  };

  return (
    <ContentWrapper>
      <div className="counter">
        {cardPile.length ? `Cards left in deck: ${cardPile.length}` : ''}
      </div>

      <Card content={currentCard} flipped={flipped} setFlipped={setFlipped} />

      <div>
        {currentCard.start ? (
          <CardButton color="#107bbd" onClick={() => randomizeDeck()}>
            Start
          </CardButton>
        ) : cardPile.length === 0 ? (
          <CardButton color="black" onClick={() => randomizeDeck()}>
            Reset
          </CardButton>
        ) : (
          <>
            <CardButton color="#C57B57" onClick={() => updateCard()}>
              Again
            </CardButton>
            <CardButton color="#70A288" onClick={() => removeCard()}>
              Good
            </CardButton>
          </>
        )}
      </div>
    </ContentWrapper>
  );
};

const mapStateToProps = state => {
  return {
    currentDeck: state.deck.currentDeck
  };
};

export default connect(mapStateToProps)(StudyCards);
