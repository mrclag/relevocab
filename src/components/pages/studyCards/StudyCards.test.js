import React from 'react';
import { StudyCards } from './StudyCards.jsx';
import { shallow } from 'enzyme';

describe('StudyCards', () => {
  const studyCards = shallow(<StudyCards />);

  it('renders correctly', () => {
    expect(studyCards).toMatchSnapshot();
  });
});
