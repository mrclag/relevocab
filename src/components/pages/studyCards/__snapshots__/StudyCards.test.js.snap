// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyCards renders correctly 1`] = `ShallowWrapper {}`;
